# NE 579 Homework Number 2: Data Statistics

## Import required libraries

In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import scipy.io as sio
from scipy.stats import trim_mean, skew, kurtosis

## Load the data from the file

In [2]:
fn = 'hwkdata.mat'
data_dict = sio.loadmat(fn)
print(data_dict.keys())

dict_keys(['__header__', '__version__', '__globals__', 'x', 'y'])


In [4]:
data_dict['x'].shape

(252, 14)

In [5]:
data_dict['y'].shape

(252, 1)

In [6]:
data = np.append(data_dict['x'], data_dict['y'], 1)

In [7]:
data.shape

(252, 15)

Now the 2-d array `data` contains 252 samples, each with 14 input variables and 1 output variable

## Calculate statistical properties of the data:
- Maximum
- Minimum
- Mean
- Median
- 20% trimmed mean
- Standard deviation
- Variance
- Skewness
- Kurtosis

### Maximum

In [9]:
np.max(data, axis=0)

array([ 81.  , 363.15,  77.75,  48.9 ,  51.2 , 136.2 , 148.1 , 147.7 ,
        87.3 ,  49.1 ,  33.9 ,  45.  ,  34.9 ,  21.4 ,  45.1 ])

### Minimum

In [10]:
np.min(data, axis=0)

array([ 22. , 118.5,  29.5,  18.1,  31.1,  79.3,  69.4,  85. ,  47.2,
        33. ,  19.1,  24.8,  21. ,  15.8,   0. ])

### Mean

In [11]:
np.mean(data, axis=0)

array([ 44.88492063, 178.92440476,  70.14880952,  25.43690476,
        37.99206349, 100.82420635,  92.55595238,  99.9047619 ,
        59.40595238,  38.59047619,  23.10238095,  32.2734127 ,
        28.66388889,  18.2297619 ,  18.93849206])

### Median

In [12]:
np.median(data, axis=0)

array([ 43.  , 176.5 ,  70.  ,  25.05,  38.  ,  99.65,  90.95,  99.3 ,
        59.  ,  38.5 ,  22.8 ,  32.05,  28.7 ,  18.3 ,  19.  ])

### 20% Trimmed Mean

In [13]:
trim_mean(data, 0.2, axis=0)

/Users/jrpowers-luhn/miniconda3/envs/579/lib/python3.6/site-packages/scipy/stats/stats.py:2831: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  return np.mean(atmp[sl], axis=axis)


array([ 44.44078947, 176.55361842,  70.25      ,  25.05986842,
        37.92894737, 100.12763158,  91.80789474,  99.32828947,
        59.13092105,  38.49144737,  22.91381579,  32.16118421,
        28.69934211,  18.21776316,  18.86578947])

### Standard Deviation

In [14]:
np.std(data, axis=0)

array([12.57701082, 29.3307901 ,  3.65558099,  3.64086527,  2.4260852 ,
        8.41373177, 10.76166054,  7.14982914,  5.2395251 ,  2.4070145 ,
        1.69152717,  3.0152732 ,  2.01667787,  0.93173074,  7.73546169])

### Variance

In [15]:
np.var(data, axis=0)

array([158.18120118, 860.29524766,  13.36327239,  13.25589994,
         5.88588939,  70.79088231, 115.81333759,  51.12005669,
        27.4526233 ,   5.79371882,   2.86126417,   9.09187248,
         4.06698964,   0.86812217,  59.83736757])

### Skewness

In [16]:
skew(data, axis=0)

array([ 0.28183067,  1.19807685, -5.35287991,  1.55239101,  0.54932508,
        0.6774921 ,  0.83341904,  1.48820106,  0.81631331,  0.51366304,
        2.24168861,  0.28382759, -0.21802506,  0.27993485,  0.14338879])

### Kurtosis

In [17]:
kurtosis(data, axis=0)

array([-0.43193939,  5.14182358, 58.34569713,  6.55632326,  2.64223799,
        0.94408639,  2.18073601,  7.30021682,  2.5894009 ,  1.01687492,
       11.68578418,  0.46494688,  0.82550054,  0.36415485, -0.32445733])

#### Also calculate $Kurt - 3$

In [18]:
kurtosis(data, axis=0) - 3 * np.ones_like(kurtosis(data, axis=0))

array([-3.43193939,  2.14182358, 55.34569713,  3.55632326, -0.35776201,
       -2.05591361, -0.81926399,  4.30021682, -0.4105991 , -1.98312508,
        8.68578418, -2.53505312, -2.17449946, -2.63584515, -3.32445733])